text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+among others can be addressed by their name.
+
+```rust
+unsafe {
+    asm!("out 0x64, {}", in("eax") cmd);
+}
+```
+
+In this example we call the `out` instruction to output the content of the `cmd` variable
+to port `0x64`. Since the `out` instruction only accepts `eax` (and its sub registers) as operand
+we had to use the `eax` constraint specifier.
+
+It is somewhat common that instructions have operands that are not explicitly listed in the
+assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:
+
+```rust
+fn mul(a: u32, b: u32) -> u64 {
+    let lo: u32;
+    let hi: u32;
+
+    unsafe {
+        asm!(
+            // The x86 mul instruction takes eax as an implicit input and writes
+            // the 64-bit result of the multiplication to eax:edx.
+            "mul {}",
+            in(reg) a, in("eax") b,
+            lateout("eax") lo, lateout("edx") hi
+        );
+    }
+
+    hi as u64 << 32 + lo as u64
+}
+```
+
+This uses the `mul` instruction to multiply two 32-bit inputs with a 64-bit result.
+The only explicit operand is a register, that we fill from the variable `a`.
+The second implicit operand is the `eax` register which we fill from the variable `b`.
+The lower 32 bits of the result are stored in `eax` from which we fill the variable `lo`.
+The higher 32 bits are stored in `edx` from which we fill the variable `hi`.
+
+Note that `lateout` must be used for `eax` here since we are specifying the same register as both an input and an output.
+
+## Clobbered registers
+
+In many cases inline assembly will modify state that is not needed as an output.
+Usually this is either because we have to use a scratch register in the assembly,
+or instructions modify state that we don't need to further examine.
+This state is generally referred to as being "clobbered".
+We need to tell the compiler about this since it may need to save and restore this state
+around the inline assembly block.
+
+```rust
+let ebx: u32;
+let ecx: u32;
+
+unsafe {
+    asm!(
+        "cpuid",
+        in("eax") 4, in("ecx") 0,
+        lateout("ebx") ebx, lateout("ecx") ecx,
+        lateout("eax") _, lateout("edx") _
+    );
+}
+
+println!(
+    "L1 Cache: {}",
+    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1) * ((ebx & 0xfff) + 1) * (ecx + 1)
+);
+```
+
+In the example above we use the `cpuid` instruction to get the L1 cache size.
+This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.
+
+However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.
+
+This can also be used with a general register class (e.g. `reg`) to obtain a scratch register for use inside the asm code.
```

Comment:
An example of doing this would be nice, just to show how it all fits together.

Reply:
I added an example.